### PR TITLE
Initial Command Line Tool Code

### PR DESCRIPTION
**Pull request by nokeeo**

### CODE REVIEW

File: cliTool/argumentParser.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+function getFormat(formatString) {
+	let format = formatString.toLowerCase();
+	if(format == 'apa' || format == 'a') {
+		return require('../core/model/formats/apa');
+	}
+	else if(format == 'chicago' || format == 'c') {
+		return require('../core/model/formats/chicago');
+	}
+
+	throw new Error(formatString + ' is an unsuported citation format. Try apa or chicago');
```
Comment (faokryn) [inline cliTool/argumentParser.js]: Might be a little clearer if we quote the arguments in the message, i.e. '' is an unsupported citation format. Try "apa" or "chicago".'

File: cliTool/argumentParser.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+ * A parse function must have the parameters:
+ * - the index that the flag occured.
+ * - the array of the args
+ * - the format option object that it must modify based on the value of the flag.
+ * The function must return the index after the last argument it has consumed.  This index does not need to be in bounds, but it should
+ * me greater than the index passed in.
+ */
+let optionFlags = {
+	'-f' : parseFormat
+}
```
Comment (faokryn) [inline cliTool/argumentParser.js]: End your file with an empty line.

File: core/citationCore.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+		}
+	}
+}
+
+/**
+ * The callback for the CitationCore generate function
+ * @callback CitationCore~generate
+ * @param {string} The Formatted citation
+ * @param {Array<Messages>}
+ */
```
Comment (faokryn) [inline core/citationCore.js]: End your file with an empty line.

File: core/citationFormatter.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+		 * Formats a citation object to a string based on set format options
+		 * @param {Citation} citation to format
+		 * @param {Array} array of format options
+		 */
+		'format' : function(citation, formatOptions) {
+			// TODO: Generate citation
+			return 'Citation'
+		}
+	}
+}()
```
Comment (faokryn) [inline core/citationFormatter.js]: End your file with an empty line.

File: core/model/formatOptions.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+ * @property {Formatter} - The formatter to use when generating the citation string. Default is APA.
+ */
+class FormatOptions {
+	constructor() {
+		this.url = '';
+		this.style = require('./formats/apa');
+	}
+}
+
+module.exports = FormatOptions;
```
Comment (faokryn) [inline core/model/formatOptions.js]: End your file with an empty line.

File: core/model/formats/apa.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+		}
+
+		// URL
+		if(sourceData.url != null) {
+			returnString += ' Retrieved From: ' + sourceData.url; 
+		}
+
+		return returnString;
+	}
+}
```
Comment (faokryn) [inline core/model/formats/apa.js]: End your file with an empty line.

File: core/model/formats/chicago.js [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+		}
+
+		// URL
+		if(sourceData.url != null) {
+			returnString += sourceData.url;
+		}
+
+		return returnString;
+	}
+}
```
Comment (faokryn) [inline core/model/formats/chicago.js]: End your file with an empty line.

File: core/sourceHandlers/gitHub.js [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+		// sourceData.url = url;
+		// sourceData.licence = 'MIT';
+		// sourceData.description = 'This is a test project';
+		// sourceData.uid = null;
+		// console.log(sourceData);
+
+		// var messages = []
+		// callback(sourceData, messages)
+	}
+};
```
Comment (faokryn) [inline core/sourceHandlers/gitHub.js]: End your file with an empty line.

File: core/urlResolver.js [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+		// See if we can match a generator
+		for(let handler of handlers) {
+			if(handler.canHandle(url)) {
+				return handler;
+			}
+		}
+
+		return null;
+	}
+}
```
Comment (faokryn) [inline core/urlResolver.js]: End your file with an empty line.

File: core/citationCore.js [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+		let sanitizedUrl = formatOptions.url.replace(/^http(s)?\:\/\//, '').replace(/^www\./, '');
+		let urlHandler = UrlResolver.getHandler(sanitizedUrl);
+		if(urlHandler != null) {
+			urlHandler.fetch(sanitizedUrl, (sourceData, messages) => {
+				let citation = formatOptions.style.format(sourceData);
+				callback(citation, [])
+			})
+		}
+		else {
+			// TODO: Populate array with error
```
Comment (faokryn) [inline core/citationCore.js]: This looks like it was done. You should remove it.

File: core/citationFormatter.js [hunk 11]
```diff
@@ -0,0 +1,13 @@
+module.exports = function() {
+	return {
+		/**
+		 * Formats a citation object to a string based on set format options
+		 * @param {Citation} citation to format
+		 * @param {Array} array of format options
+		 */
+		'format' : function(citation, formatOptions) {
+			// TODO: Generate citation
+			return 'Citation'
```
Comment (faokryn) [inline core/citationFormatter.js]: Is this the right file? Should this be doing something?

Comment (faokryn) [reply core/citationFormatter.js]: I can't find anywhere this is used at all? Is it a dead file?

File: core/model/formats/apa.js [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+	 * Formats a source data object to a citation
+	 * @param {SourceData} sourceData - The SourceData object to generate the citation string from
+	 * @return {string} - The citation in APA format
+	 */
+	format : (sourceData) => {
+		let returnString = '';
+		
+		// Authors
+		let numAuthors = sourceData.authors.length;
+		sourceData.authors.forEach((author, index, array) => {
```
Comment (faokryn) [inline core/model/formats/apa.js:40]: We should eventually have code to add "et al" or however APA does large numbers of authors

File: core/citationFormatter.js (hunk 11 above)
Comment (nokeeo) [reply core/citationFormatter.js]: Dead and deleted

File: core/model/formats/apa.js (hunk 12 above)
Comment (nokeeo) [reply core/model/formats/apa.js:40]: Agreed.  Release Year should also be n.d. If a release date does not exist. We should consider for future releases

**Review by sam1360 — APPROVED** [review]
Looks solid.

**Comment by faokryn** [comment]
:+1: LGTM: there are a few places we may need to make some changes in the future, but since this is for an initial prototype to get that discussion started I think it's fine.

**Comment by faokryn** [comment]
@nokeeo ಠ_ಠ

**Pull request merged**